### PR TITLE
Add support for pandas.DataFrame and list in `fit`

### PR DESCRIPTION
**Pull request by IncubatorShokuhou**
Now, the `fit` method only support `np.array` for input. However, most ml algorithms with scikit-learn-Compatible API (i.e. [XGBoost](https://github.com/dmlc/xgboost) , [NGBoost](https://github.com/stanfordmlgroup/ngboost)) support `DataFrame` object or List or numpy array of predictors (n x p) in numeric format using `sklearn.utils.check_array` .
This RP is to make `Deep-Forest` consistent with other algorithms so that it will be more easy to be used in other integrated machine learning frameworks (i.e. [PyCaret](https://github.com/pycaret/pycaret)).

### CODE REVIEW

**Comment by xuyxu** [comment]
Hi @IncubatorShokuhou , thanks for the PR! I will take a look soon.

**Comment by IncubatorShokuhou** [comment]
@xuyxu  Sorry, but I just find out that `y = check_array(y)` is added by mistake. I'll change is ASAP.

**Comment by xuyxu** [comment]
In order to pass the code quality CI, you could use the command `black --skip-string-normalization --config pyproject.toml ./` in the project directory to automatically re-format the code.

File: deepforest/cascade.py [hunk 1]
```diff
@@ -762,7 +763,8 @@ def n_aug_features_(self):
 
     # flake8: noqa: E501
     def fit(self, X, y, sample_weight=None):
-
+        X = check_array(X)
```
Comment (xuyxu) [inline deepforest/cascade.py:766]: So `check_array` from scikit-learn will automatically cast the dataframe into the numpy array?

**Comment by IncubatorShokuhou** [comment]
> 
> 
> In order to pass the code quality CI, you could use the command `black --skip-string-normalization --config pyproject.toml ./` in the project directory to automatically re-format the code.

@xuyxu Maybe tomorrow :)  I cannot access to my work computer right now.

File: deepforest/cascade.py (hunk 1 above)
Comment (IncubatorShokuhou) [reply deepforest/cascade.py:766]: Yes. A similar method is used in https://github.com/stanfordmlgroup/ngboost/blob/master/ngboost/api.py#300, and I believe it should be right.

Comment (xuyxu) [reply deepforest/cascade.py:766]: I will try out the model built on this PR when I get a moment. Thanks!

**Comment by xuyxu** [comment]
@all-contributors please add @IncubatorShokuhou  for code

**Review by xuyxu — COMMENTED** [review]
LGTM.

**Comment by IncubatorShokuhou** [comment]
I just noticed that `check_array` is also needed in `predict` et al. Will soon fix it.

**Comment by IncubatorShokuhou** [comment]
> I just noticed that `check_array` is also needed in `predict` et al. Will soon fix it.

Fixed.

**Comment by xuyxu** [comment]
@all-contributors please add @IncubatorShokuhou  for code

File: deepforest/cascade.py [hunk 2]
```diff
@@ -315,7 +316,7 @@ def _build_regressor_predictor(
 
     Parameters
     ----------
-    X : :obj:`numpy.ndarray` of shape (n_samples, n_features)
+    X : :obj:`numpy.ndarray` or `pandas.DataFrame` or List of shape (n_samples, n_features)
```
Comment (xuyxu) [inline deepforest/cascade.py:319]: I think there is no need to further add `or List of shape (n_samples, n_features)`

**Review by xuyxu — COMMENTED** [review]
See comments above.

**Comment by xuyxu** [comment]
I think it would be better if we further accept `y` as a pandas dataframe ;-)

For now, the `CascadeForestRegressor` corrupts with the following code snippet:

``` python
import pandas as pd
from sklearn.datasets import load_boston
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error

from deepforest import CascadeForestRegressor

X, y = load_boston(return_X_y=True)
X, y = pd.DataFrame(X), pd.DataFrame(y)
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=1)
model = CascadeForestRegressor(random_state=1)
model.fit(X_train, y_train)
y_pred = model.predict(X_test)
mse = mean_squared_error(y_test, y_pred)
print("\nTesting MSE: {:.3f}".format(mse))
```

**Comment by IncubatorShokuhou** [comment]
I have tested that if I use `check_array` to `y`, there will be an error. Maybe I can do some further test to see what to do to make it more consistent&nbsp;with lgb/xgb et. al.

---Original---
From: "Yi-Xuan ***@***.***&gt;
Date: Fri, Jul 16, 2021 20:32 PM
To: ***@***.***&gt;;
Cc: "Hao ***@***.******@***.***&gt;;
Subject: Re: [LAMDA-NJU/Deep-Forest] Add support for pandas.DataFrame and list in `fit` (#86)




 
I think it would be better if we further accept y as a pandas dataframe ;-)
 
For now, the CascadeForestRegressor corrupts with the following code snippet:
 import pandas as pd from sklearn.datasets import load_boston from sklearn.model_selection import train_test_split from sklearn.metrics import mean_squared_error from deepforest import CascadeForestRegressor X, y = load_boston(return_X_y=True) X, y = pd.DataFrame(X), pd.DataFrame(y) X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=1) model = CascadeForestRegressor(random_state=1) model.fit(X_train, y_train) y_pred = model.predict(X_test) mse = mean_squared_error(y_test, y_pred) print("\nTesting MSE: {:.3f}".format(mse))

**Comment by xuyxu** [comment]
Let me know if you need any help

**Comment by IncubatorShokuhou** [comment]
Now the processing method of `X` and `y` is modified in `fit` using `check_X_y` referring to lightgbm.  Now all array-like can be correctly handled.
And the code comment is modified (`array-like`) referring to xgboost and lightgbm.

File: deepforest/cascade.py (hunk 2 above)
Comment (IncubatorShokuhou) [reply deepforest/cascade.py]: Now the code comment is modified to `array-like` referring to xgboost and lightgbm.

**Comment by xuyxu** [comment]
It looks like the CI failed, possibly because `check_X_y` cannot properly handle multi-output regression targets.

Could you further check it? You can find unit test [here](https://github.com/LAMDA-NJU/Deep-Forest/blob/master/tests/test_set_custom_estimator.py#L94), thanks.

**Comment by IncubatorShokuhou** [comment]
> It looks like the CI failed, possibly because `check_X_y` cannot properly handle multi-output regression targets.
> 
> Could you further check it? You can find unit test [here](https://github.com/LAMDA-NJU/Deep-Forest/blob/master/tests/test_set_custom_estimator.py#L94), thanks.

@xuyxu  Fixed using `multi_output=True if type_of_target(y) in ("continuous-multioutput", "multiclass-multioutput") else False` in `check_X_y`.

**Review by xuyxu — COMMENTED** [review]
LGTM.

**Comment by xuyxu** [comment]
Merged, thanks!
